File: ocaml/gup/builder.ml
open Std
open Path
open Error
module Log = (val Var.log_module "gup.builder")

let _guess_executable path =
	Lwt_io.with_file ~mode:Lwt_io.input path (fun file ->
		Lwt_io.read ~count:255 file |> Lwt.map (fun initial ->
			if CCString.prefix ~pre:"#!" initial then (
				let initial = CCString.drop 2 initial in
				let (line, _) =
					CCString.Split.left ~by:"\n" initial |> CCOpt.get_or ~default:(initial,"")
				in
				let args = Str.split (Str.regexp "[ \t]+") line in
				match args with
					| bin :: rest ->
						let bin = if CCString.prefix ~pre:"." bin
							then Filename.concat (Filename.dirname path) bin
							else bin
						in
						if Util.is_absolute bin && not (Sys.file_exists bin) then (
							(* special-case: we ignore /path/to/<env> for compatibility with plain shell scripts on weird platforms *)
							if Filename.basename bin = "env" then rest
							else Error.raise_safe "No such interpreter: %s" bin;
						) else bin :: rest
					| [] -> []
			) else []
		)
	)

let in_dir wd action =
	let initial_wd = Lazy.force Var_global.cwd in
	Unix.chdir wd;
	Util.finally_do (fun _ -> Unix.chdir initial_wd) () action

let perform_build ~lease ~var ~toplevel (buildable: Buildable.t) = (
	(* actually perform a build *)
	let path = Buildable.target_path buildable in
	let state = new State.target_state ~var path in
	let exe_path = Buildable.script buildable in
	if not (Absolute.exists exe_path) then
		Error.raise_safe "Build script does not exist: %s" (Absolute.to_string exe_path);

	state#perform_build buildable (fun deps ->
		let path_str = ConcreteBase.to_string path in
		let target = Buildable.target buildable in
		let basedir = target |> RelativeFrom.base in
		let basedir_str = Concrete.to_string basedir in
		Util.makedirs basedir_str;

		let env = Unix.environment_map ()
			|> EnvironmentMap.add "GUP_TARGET" path_str
			|> Parallel.Jobpool.extend_env ~lease
			|> Var.extend_env var
			|> EnvironmentMap.array
		in

		let relative_to_root_cwd : ConcreteBase.t -> string = fun path ->
			ConcreteBase.rebase_to Var.root_cwd path
			|> RelativeFrom.relative |> Relative.to_string
		in

		let target_relstr = relative_to_root_cwd path in

		let output_file = Concrete.resolve_abs (state#meta_path "out") |> Concrete.to_string in
		Util.try_remove output_file;
		let cleanup_output_file = ref true in
		let cleanup () =
			if !cleanup_output_file then
				Util.try_remove output_file
		in

		let do_build () =
			Log.info var (fun m->m"%s" target_relstr);
			let exe_path_str = Absolute.to_string exe_path in
			let%lwt (mtime, executable) = Util.lwt_zip
				(Util.get_mtime path_str)
				(_guess_executable exe_path_str) in
			let args = List.concat
				[
					executable;
					[ exe_path_str; output_file;
						target |> RelativeFrom.relative |> Relative.to_string
					]
				]
			in

			if !Var_global.trace then begin
				Log.info var (fun m->m " # %s" basedir_str);
				Log.info var (fun m->m " + %a" PP.(list string) args)
			end else begin
				Log.trace var (fun m->m " from cwd: %s" basedir_str);
				Log.trace var (fun m->m "executing: %a" PP.(list string) args)
			end;

			let%lwt ret = try%lwt in_dir basedir_str (fun () -> Lwt_process.exec
					~env:env
					((List.hd args), (Array.of_list args))
				)
				with ex -> begin
					Log.err var (fun m -> m"%s is not executable and has no shebang line" exe_path_str);
					raise ex
				end
			in
			let%lwt new_mtime = Util.get_mtime path_str in
			let target_changed = neq (CCOpt.compare Big_int.compare_big_int) mtime new_mtime in
			let%lwt () = if target_changed then (
				let p = CCOpt.pp Util.big_int_pp in
				Log.trace var (fun m->m "old_mtime=%a, new_mtime=%a" p mtime p new_mtime);
				if (Util.lisdir path_str) then Lwt.return_unit else (
					(* directories often need to be created directly *)
					let expect_clobber = match deps with None -> false | Some d -> d#clobbers in
					if (toplevel || (not expect_clobber)) then (
						Log.warn var (fun m->m "%s modified %s directly"
							(relative_to_root_cwd (ConcreteBase.resolve_abs exe_path))
							target_relstr)
					);
					state#mark_clobbers
				)
			) else Lwt.return_unit in

			match ret with
				| Unix.WEXITED 0 -> begin
					let%lwt () = if Util.lexists output_file then (
						(* If both old and new exist, and either is a directory,
						 * remove the old dir before renaming *)
						if (Util.lexists path_str &&
							(Util.lisdir path_str || Util.lisdir output_file)
						) then (
							Log.trace var (fun m->m "removing previous %s" path_str);
							Util.rmtree path_str
						);
						Log.trace var (fun m->m "renaming %s -> %s" output_file path_str);
						Lwt_unix.rename output_file path_str
					) else (
						Log.trace var (fun m->m "output file %s did not get created" output_file);
						if (not target_changed) && (Util.lexists path_str) && (not (Util.islink path_str)) then (
							if Util.lexists path_str; then (
								Log.warn var (fun m->m "Removing stale target: %s" target_relstr)
							);
							(* TODO make this an lwt.t *)
							Util.try_remove path_str;
							Lwt.return_unit
						) else Lwt.return_unit
					) in
					cleanup_output_file := false; (* not needed *)
					Lwt.return true
				end
				| Unix.WEXITED code -> begin
					Log.trace var (fun m->m "builder exited with status %d" code);
					let temp_file = if Var_global.keep_failed_outputs () && Util.lexists output_file
						then (
							cleanup_output_file := false; (* not wanted *)
							Some (relative_to_root_cwd (ConcreteBase.resolve ~cwd:var.Var.cwd output_file))
						)
						else None
					in
					raise @@ Target_failed (target_relstr, Some code, temp_file)
				end
				| _ -> begin
					Log.trace var (fun m->m "builder was terminated");
					raise @@ Target_failed (target_relstr, None, None)
				end
		in
		(
			try%lwt do_build ()
			with e -> raise e
		) [%lwt.finally
			Lwt.return (cleanup ())
		]
	)
)

let _build_if_dirty ~lease ~var ~cache ~dry = (
	(*
	* Returns whether the dependency was built.
	* If `dry` is true, does not build but returns
	* whether any target would be built.
	*
	* Awkwardly, there are two different forms of recursion in targets:
	* - buildscript recursion, e.g. when target `foo` is built by
	*   `foo-builder`, but `foo-builder` is itself built by
	*   `foo-builder-builder`, etc.
	*   The entire chain of buildscript recursion is returned by
	*   the search for the initial buildscript for a target.
	*
	* - runtime dependencies, e.g. anything you call `gup -u` on
	*   during the build process. These are stored in the deps file
	*   within the gup directory, so they aren't known up front.
	*
	* `build_target_if_dirty` handles the core building logic, and
	* that's invoked via either `build_recursive_if_dirty` (for buildscript
	* recursion) or `build_path_if_dirty` (for runtime dependencies)
	*)
	let perform_build = if dry
		then (fun _ -> Lwt.return_true)
		else perform_build ~lease ~var ~toplevel:false in

	(* builds a single buildable if its state is dirty *)
	let rec build_target_if_dirty buildable = (
		let target_path = Buildable.target_path buildable in
		PathMap.cached cache target_path (fun () ->
			Log.debug var (fun m->m "%s [check]"
				(ConcreteBase.rebase_to Var.root_cwd target_path
				|> RelativeFrom.relative |> Relative.to_string)
			);
			let state = new State.target_state ~var target_path in
			Log.trace var (fun m->m "checking whether %s is dirty" state#path_repr);
			let%lwt deps = state#deps in
			match deps with
				| None -> (
					Log.debug var (fun m->m "DIRTY: %s (is buildable but has no stored deps)" state#path_repr);
					perform_build buildable
				)
				| Some deps -> (
					if deps#already_built then (
						Log.trace var (fun m->m "CLEAN: %s has already been built in this invocation" state#path_repr);
						Lwt.return_false
					) else (
						if dry then (
							(* In order to determine whether this file _may_ be dirty,
							 * we inject a fake build function which just sets a flag
							 * if any possibly-dirty dependency needs a build. *)
							let open Lwt in
							let child_dirty = ref false in
							let%lwt dirty = deps#is_dirty ~var buildable (fun path ->
								let%lwt child_built = build_path_if_dirty path in
								if child_built then child_dirty := true;
								return_false (* short-circuit any further checking *)
							) in
							return (dirty || !child_dirty)
						)
						else (
							Lwt.bind (deps#is_dirty ~var buildable build_path_if_dirty) (function
								| true -> perform_build buildable
								| false -> Lwt.return_false
							)
						)
					)
				)
			)
	)

	(* builds a single _path_ if dirty (used by deps#is_dirty) *)
	and build_path_if_dirty path = (
		let path = ConcreteBase.resolve_relfrom path in
		let%lwt builder = Gupfile.find_builder ~var path in
		match builder with
			| None ->
				Log.trace var (fun m->m "CLEAN: %s (not a target)" (ConcreteBase.to_string path));
				Lwt.return_false
			| Some buildable -> build_recursive_if_dirty buildable
	)

	and build_recursive_if_dirty (buildable: Buildable.t Recursive.t) = (
		Log.trace var (fun m->m "checking if %a needs to be built" (Recursive.print Buildable.print) buildable);
		match buildable with
			| Recursive.Recurse { parent; child = self } ->
				Lwt.bind (build_recursive_if_dirty parent) (function
					| true -> perform_build self
					| false -> build_target_if_dirty self
				)
			| Recursive.Terminal self ->
				build_target_if_dirty self
	) in

	build_recursive_if_dirty
)

let is_dirty ~var buildable : bool Lwt.t =
	_build_if_dirty ~lease:None ~var ~cache:(ref PathMap.empty) ~dry:true buildable

let _build_cache = ref PathMap.empty

let rec build ~lease ~var ~update (buildable: Buildable.t Recursive.t) : bool Lwt.t =
	if update then (
		_build_if_dirty ~lease ~var ~cache:_build_cache ~dry:false buildable
	) else (
		match buildable with
			| Recursive.Recurse { parent; child } ->
				Lwt.bind (build ~lease ~var ~update:true parent) (fun (_:bool) ->
					perform_build ~lease ~var ~toplevel:true child
				)
			| Recursive.Terminal child -> perform_build ~lease ~var ~toplevel:true child
	)

type prepared_build = [
	| `Target of Buildable.t Recursive.t
	| `Symlink_to of RelativeFrom.t
]

let prepare_build ~var (path: ConcreteBase.t) : prepared_build option Lwt.t =
	Gupfile.find_builder ~var path |> Lwt.map (function
		| Some buildable -> Some (`Target buildable)
		| None -> (match ConcreteBase.readlink path with
			(* # this target isn't buildable, but its symlink destination might be *)
			| `concrete _ -> None
			| `link dest -> Some (`Symlink_to dest)
		)
	)

File: python/gup/builder.py
from __future__ import print_function
import os
from os import path
import errno
import subprocess
import logging

from .gupfile import Builder
from .error import *
from .util import *
from .state import TargetState
from .log import getLogger
from .var import ROOT_CWD, XTRACE, IS_WINDOWS, keep_failed_outputs
from .path import resolve_base
from .parallel import extend_build_env
_log = getLogger(__name__)

try:
	from pipes import quote
except ImportError:
	from shlex import quote

def prepare_build(p):
	p = resolve_base(p)
	builder = Builder.for_target(p)
	_log.trace('prepare_build(%r) -> %r' % (p, builder))
	if builder is not None:
		return Target(builder)
	return None

def _build_if_dirty(target, allow_build):
	'''
	Returns whether the dependency was built (or would be).
	Also builds any targets required to check dirtiness.
	'''

	def perform_build(target):
		if allow_build:
			target.perform_build(from_update=True)
		return True

	def build_target_if_dirty(target):
		if target.builder.parent is not None:
			if build_target_if_dirty(Target(target.builder.parent)):
				_log.debug("DIRTY: builder was rebuilt")
				return perform_build(target)

		deps = target.state.deps()
		if deps is None:
			_log.debug("DIRTY: %s (is buildable but has no stored deps)", target.path)
			return perform_build(target)

		if deps.already_built():
			_log.trace("DIRTY: %s was previously built in this invocation", target.path)
			return True

		built_children = set()
		def build_child_if_dirty(path):
			if path in built_children:
				return False
			else:
				built_children.add(path)
				_log.trace("Recursing over dependency: %s -> %s", target.path, path)
				child = prepare_build(path)
				if child is not None:
					child_dirty = build_target_if_dirty(child)
					_log.trace("_is_dirty(%s) -> %s", child, child_dirty)
					return child_dirty
				return False

		if not allow_build:
			child_dirty = []
			def wrapped_build(path):
				built = build_child_if_dirty(path)
				if built:
					child_dirty.append(True)
				return False
			dirty = deps.is_dirty(target.builder, wrapped_build)
			dirty = dirty or bool(child_dirty)
		else:
			dirty = deps.is_dirty(target.builder, build_child_if_dirty)
		_log.trace("deps.is_dirty(%r) -> %r", target.path, dirty)
		if dirty:
			return perform_build(target)
		return False

	return build_target_if_dirty(target)

class Target(object):
	def __init__(self, builder):
		self.builder = builder
		self.path = self.builder.target_path
		self.state = TargetState(self.path)
	
	def __repr__(self):
		return 'Target(%r)' % (self.path,)
	
	def build_or_update(self, update):
		if update:
			built = _build_if_dirty(self, True)
			if not built:
				_log.trace("no build needed")
		else:
			self.perform_build(False)
			built = True

		return built

	def is_dirty(self):
		return _build_if_dirty(self, allow_build = False)

	def perform_build(self, from_update):
		self.state.perform_build(self.builder, lambda deps: self._perform_build(deps, from_update))

	def _perform_build(self, deps, from_update):
		'''
		Assumes locks are held (by state.perform_build)
		'''
		assert self.builder is not None
		assert os.path.exists(self.builder.path)
		exe_path = path.abspath(self.builder.path)
		exe_path_relative_to_cwd = os.path.relpath(exe_path,ROOT_CWD)

		# dest may not exist, if a /gup/ directory is in use
		basedir = self.builder.basedir
		mkdirp(basedir)

		env = os.environ.copy()
		env['GUP_TARGET'] = os.path.abspath(self.path)
		extend_build_env(env)

		target_relative_to_cwd = os.path.relpath(self.path, ROOT_CWD)

		output_file = os.path.abspath(self.state.meta_path('out'))
		try_remove(output_file)

		cleanup_output_file = True
		try:
			args = [exe_path, output_file, self.builder.target]
			_log.info(target_relative_to_cwd)
			mtime = get_mtime(self.path)

			exe = _guess_executable(exe_path)

			if exe is not None:
				args = exe + args

			if XTRACE:
				_log.info(' # %s'% (os.path.abspath(basedir),))
				_log.info(' + ' + ' '.join(map(quote, args)))
			else:
				_log.trace(' from cwd: %s'% (os.path.abspath(basedir),))
				_log.trace('executing: ' + ' '.join(map(quote, args)))

			try:
				ret = self._run_process(args, cwd = basedir, env = env)
			except OSError:
				if exe: raise # we only expect errors when we could deduce no executable
				raise SafeError("%s is not executable and has no shebang line" % (exe_path_relative_to_cwd))

			new_mtime = get_mtime(self.path)
			target_changed = mtime != new_mtime
			if target_changed:
				_log.trace("old_mtime=%r, new_mtime=%r" % (mtime, new_mtime))
				if not lisdir(self.path):
					# directories often need to be created directly
					self.state.mark_clobbers()
					expect_clobber = False if deps is None else deps.clobbers
					if not (from_update and expect_clobber):
						_log.warning("%s modified %s directly" % (exe_path_relative_to_cwd, self.path))
			if ret == 0:
				if os.path.lexists(output_file):
					if os.path.lexists(self.path) and (
						lisdir(self.path) or lisdir(output_file)
					):
						_log.trace("removing previous %s", self.path)
						try_remove(self.path)
					rename(output_file, self.path)
				else:
					if (not target_changed) and (os.path.lexists(self.path)) and (not os.path.islink(self.path)):
						_log.warning("Removing stale target: %s", target_relative_to_cwd)
						try_remove(self.path)
				cleanup_output_file = False # not needed
			else:
				temp_file = None
				if keep_failed_outputs():
					cleanup_output_file = False # not wanted
					if os.path.lexists(output_file):
						temp_file = os.path.relpath(output_file, ROOT_CWD)
				_log.trace("builder exited with status %s" % (ret,))
				raise TargetFailed(target_relative_to_cwd, ret, temp_file)
		finally:
			if cleanup_output_file:
				try_remove(output_file)
		return True

	
	def _run_process(self, args, cwd, env):
		try:
			proc = subprocess.Popen(args, cwd = cwd, env = env, close_fds=False)
		except OSError as e:
			if e.errno == errno.ENOENT:
				raise SafeError("Executable not found: %s" % (args[0],))
			raise e
		return proc.wait()

def _guess_executable(p):
	with open(p, 'rb') as f:
		line = f.readline(255)
	if not line.startswith(b'#!'):
		return None
	args = line[2:].decode('ascii').strip().split()
	if not args: return None

	bin = args[0]
	if bin.startswith('.'):
		# resolve relative paths relative to containing dir
		bin = args[0] = os.path.join(os.path.dirname(p), args[0])
	if IS_WINDOWS:
		bin = _resolve_windows_binary(bin)

	if os.path.isabs(bin) and not os.path.exists(bin):
		if os.path.basename(bin) == 'env':
			# special-cased for compatibility
			return args[1:]
		raise SafeError("No such interpreter: %s" % (os.path.abspath(bin),))

	args[0] = bin
	return args

def _resolve_windows_binary(name):
	exts = os.environ.get('PATHEXT', '').split(os.pathsep)
	def possible_file_extensions(path):
		for ext in exts:
			yield path + ext
		yield path

	def possible_paths():
		if os.path.isabs(name):
			for path in possible_file_extensions(name):
				yield path
		else:
			for prefix in os.environ['PATH'].split(os.pathsep):
				for path in possible_file_extensions(os.path.join(prefix, name)):
					yield path
	
	for path in possible_paths():
		if os.path.isfile(path) and os.access(path, os.X_OK):
			return path

	# If we found nothing, just return the original name.
	# It's probably not going to work, but Windows does
	# some nutty stuff with the registry.
	return name
